### PR TITLE
chore(experiment): Add pod network duplication experiment in generic experiments of LitmusChaos

### PR DESCRIPTION
**Pull request by uditgaurav**
Signed-off-by: Udit Gaurav <udit.gaurav@mayadata.io>

**This PR contains:** 

 - Addition a pod network duplication experiment in the generic experiment of litmus. Litmus already has a number of pod level network chaos experiments like pod network latency, loss, corruption. This will help to increase the coverage of network experiments.
- The experiment should be in go, so the PR will go to the litmus-go repository. This experiment uses netem command to generate a network duplication same as pod network latency and others.
- This issue will track the addition of the experiment. This experiment will be supported for the docker runtime environment only as of now, support for others will be added soon.

**Issues:**
- fixes: https://github.com/litmuschaos/litmus/issues/1585

**Note For Reviewer:**

### CODE REVIEW

**Comment by uditgaurav** [comment]
- **Logs:**
<details>
  <summary>Pod Network Duplication</summary>
<code>
W0622 19:09:50.291206       1 client_config.go:541] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
time="2020-06-22T19:09:50Z" level=info msg="[PreReq]: Getting the ENV for the  experiment"
time="2020-06-22T19:09:50Z" level=info msg="[PreReq]: Updating the chaos result of pod-network-duplication experiment (SOT)"
time="2020-06-22T19:09:50Z" level=info msg="The application informations are as follows\n" Ramp Time=0 Namespace=default Label="run=nginx"
time="2020-06-22T19:09:50Z" level=info msg="[Status]: Verify that the AUT (Application Under Test) is running (pre-chaos)"
time="2020-06-22T19:09:50Z" level=info msg="[Status]: Checking whether application pods are in running state"
time="2020-06-22T19:09:50Z" level=info msg="The running status of Pods are as follows" Pod=nginx-7db9fccd9b-jv85q Status=Running
time="2020-06-22T19:09:50Z" level=info msg="The running status of Pods are as follows" Pod=nginx-7db9fccd9b-wxpzv Status=Running
time="2020-06-22T19:09:50Z" level=info msg="The running status of Pods are as follows" Status=Running Pod=nginx-7db9fccd9b-xmz77
time="2020-06-22T19:09:50Z" level=info msg="[Status]: Checking whether application containers are in running state"
time="2020-06-22T19:09:50Z" level=info msg="The running status of container are as follows" container=nginx Pod=nginx-7db9fccd9b-jv85q Status=Running
time="2020-06-22T19:09:50Z" level=info msg="The running status of container are as follows" Status=Running container=nginx Pod=nginx-7db9fccd9b-wxpzv
time="2020-06-22T19:09:50Z" level=info msg="The running status of container are as follows" Status=Running container=nginx Pod=nginx-7db9fccd9b-xmz77
time="2020-06-22T19:09:50Z" level=info msg="[Prepare]: Application pod name under chaos: nginx-7db9fccd9b-jv85q"
time="2020-06-22T19:09:50Z" level=info msg="[Prepare]: Application node name: gke-cluster-1-default-pool-8774fd9c-v4br"
time="2020-06-22T19:09:50Z" level=info msg="[Prepare]: Target container name: nginx"
time="2020-06-22T19:09:50Z" level=info msg="[Status]: Checking the status of the helper pod"
time="2020-06-22T19:09:50Z" level=info msg="[Status]: Checking whether application pods are in running state"
time="2020-06-22T19:09:54Z" level=info msg="The running status of Pods are as follows" Pod=pumba-netem-vbyjby Status=Running
time="2020-06-22T19:09:54Z" level=info msg="[Status]: Checking whether application containers are in running state"
time="2020-06-22T19:09:54Z" level=info msg="The running status of container are as follows" container=pumba Pod=pumba-netem-vbyjby Status=Running
time="2020-06-22T19:09:54Z" level=info msg="[Wait]: waiting till the completion of the helper pod"
time="2020-06-22T19:09:54Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:09:55Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:09:56Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:09:57Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:09:58Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:09:59Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:10:00Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:10:01Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:10:02Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:10:03Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:10:04Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:10:05Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:10:06Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:10:07Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:10:08Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:10:09Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:10:10Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:10:11Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:10:12Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:10:13Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:10:14Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:10:15Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:10:16Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:10:17Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:10:18Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:10:19Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:10:20Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:10:21Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:10:22Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:10:23Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:10:24Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:10:25Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:10:26Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:10:27Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:10:28Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:10:29Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:10:30Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:10:31Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:10:32Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:10:33Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:10:34Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:10:35Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:10:36Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:10:37Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:10:38Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:10:39Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:10:40Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:10:41Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:10:42Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:10:43Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:10:44Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:10:45Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:10:46Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:10:47Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:10:48Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:10:49Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:10:50Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:10:51Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:10:52Z" level=info msg="helper pod status: Running"
time="2020-06-22T19:10:53Z" level=info msg="helper pod status: Succeeded"
time="2020-06-22T19:10:53Z" level=info msg="The running status of Pods are as follows" Pod=pumba-netem-vbyjby Status=Succeeded
time="2020-06-22T19:10:53Z" level=info msg="[Cleanup]: Deleting the helper pod"
time="2020-06-22T19:10:53Z" level=info msg="[Confirmation]: The app pod network duplication"
time="2020-06-22T19:10:53Z" level=info msg="[Status]: Verify that the AUT (Application Under Test) is running (post-chaos)"
time="2020-06-22T19:10:53Z" level=info msg="[Status]: Checking whether application pods are in running state"
time="2020-06-22T19:10:53Z" level=info msg="The running status of Pods are as follows" Pod=nginx-7db9fccd9b-jv85q Status=Running
time="2020-06-22T19:10:53Z" level=info msg="The running status of Pods are as follows" Pod=nginx-7db9fccd9b-wxpzv Status=Running
time="2020-06-22T19:10:53Z" level=info msg="The running status of Pods are as follows" Pod=nginx-7db9fccd9b-xmz77 Status=Running
time="2020-06-22T19:10:53Z" level=info msg="[Status]: Checking whether application containers are in running state"
time="2020-06-22T19:10:53Z" level=info msg="The running status of container are as follows" container=nginx Pod=nginx-7db9fccd9b-jv85q Status=Running
time="2020-06-22T19:10:53Z" level=info msg="The running status of container are as follows" container=nginx Pod=nginx-7db9fccd9b-wxpzv Status=Running
time="2020-06-22T19:10:53Z" level=info msg="The running status of container are as follows" container=nginx Pod=nginx-7db9fccd9b-xmz77 Status=Running
time="2020-06-22T19:10:53Z" level=info msg="[The End]: Updating the chaos result of pod-network-duplication experiment (EOT)"
</code>
</details>

File: chaoslib/pumba/network_chaos/pod-network-duplication.go [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+	if err != nil {
+		errors.Errorf("helper pod is not in running state, err: %v", err)
+	}
+
+	// Wait till the completion of helper pod
+	log.Info("[Wait]: waiting till the completion of the helper pod")
+	var endTime <-chan time.Time
+	timeDelay := time.Duration(experimentsDetails.ChaosDuration+30) * time.Second
+
+	// signChan channel is used to transmit signal notifications.
```
Comment (uditgaurav) [inline chaoslib/pumba/network_chaos/pod-network-duplication.go:83]: .

**Comment by ksatchit** [comment]
Please resolve conflicts based on previous merge.

**Comment by uditgaurav** [comment]
**Pod Network Duplication with Packet Duplication of 100%** 

Screenshot from 2020-07-02 12-17-39

**Pod Network Duplication with Packet Duplication of 50%** 

Screenshot from 2020-07-02 13-24-15

File: experiments/generic/pod-network-duplication/README.md [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+          - "chaosexperiments"
+          - "chaosresults"
+        verbs:
+          - "create"
+          - "delete"
+          - "list"
+          - "patch"
+          - "update"
+          - "get"
+    image: "litmuschaos/litmus-go:ci"
```
Comment (ispeakc0de) [inline experiments/generic/pod-network-duplication/README.md:49]: go-runner?

File: experiments/generic/pod-network-duplication/pod-network-duplication.go [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+	"github.com/litmuschaos/litmus-go/pkg/types"
+	"github.com/sirupsen/logrus"
+	"k8s.io/klog"
+)
+
+func init() {
+	// Log as JSON instead of the default ASCII formatter.
+	logrus.SetFormatter(&logrus.TextFormatter{
+		FullTimestamp: true,
+		// ForceColors:            true,
```
Comment (ispeakc0de) [inline experiments/generic/pod-network-duplication/pod-network-duplication.go:21]: Is this required?

File: chaoslib/pumba/network_chaos/pod-network-duplication.go [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+
+	//Select application pod for pod network duplication
+	appName, appNodeName, err := GetApplicationPod(experimentsDetails, clients)
+	if err != nil {
+		return errors.Errorf("Unable to get the application name and application nodename due to, err: %v", err)
+	}
+	log.Infof("[Prepare]: Application pod name under chaos: %v", appName)
+	log.Infof("[Prepare]: Application node name: %v", appNodeName)
+
+	//Get the target contaien name of the application pod
```
Comment (ispeakc0de) [inline chaoslib/pumba/network_chaos/pod-network-duplication.go:32]: contaien typo?

File: chaoslib/pumba/network_chaos/pod-network-duplication.go [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+	if experimentsDetails.TargetContainer == "" {
+		experimentsDetails.TargetContainer, err = GetTargetContainer(experimentsDetails, appName, clients)
+		if err != nil {
+			return errors.Errorf("Unable to get the target container name due to, err: %v", err)
+		}
+		log.Infof("[Prepare]: Target container name: %v", experimentsDetails.TargetContainer)
+	} else {
+		log.Infof("[Prepare]: Target container name: %v", experimentsDetails.TargetContainer)
+	}
```
Comment (ispeakc0de) [inline chaoslib/pumba/network_chaos/pod-network-duplication.go:41]:
```if experimentsDetails.TargetContainer == "" {
		experimentsDetails.TargetContainer, err = GetTargetContainer(experimentsDetails, appName, clients)
		if err != nil {
			return errors.Errorf("Unable to get the target container name due to, err: %v", err)
		}
	} 
		log.Infof("[Prepare]: Target container name: %v", experimentsDetails.TargetContainer)
```

would you like to reduce branch statements?

File: experiments/generic/pod-network-duplication/pod-network-duplication.go [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+	chaosDetails := types.ChaosDetails{}
+	eventsDetails := types.EventDetails{}
+	clients := environment.ClientSets{}
+
+	//Getting kubeConfig and Generate ClientSets
+	if err := clients.GenerateClientSetFromKubeConfig(); err != nil {
+		log.Fatalf("Unable to Get the kubeconfig due to %v", err)
+	}
+
+	//Fetching all the ENV passed for the runner pod
```
Comment (ispeakc0de) [inline experiments/generic/pod-network-duplication/pod-network-duplication.go:41]: from?

**Pull request merged**